scripts/enumerate_load_table_lookup.py

```diff
@@ -0,0 +1,377 @@
+    )
+    return output
+
+
+@timed_info
+def save_file(df, filepath, n_files=None, repartition_by=None):
+    """
+    n_files: number of target sharded files
+    repartition_by: col to repartition by
+
+    Note:
+        - Not available for load_data_lookup: df.write.partitionBy().bucketBy()
+        - df.coalesce(n): combine without shuffling, will not go larger than current_n_files
+        - df.repartition(n): shuffle and try to evenly distribute, if n > # of unique rows, some partitions will be empty
+        - df.repartition(col): shuffle and create partitions by unique record in col + 1 empty/very small partition
+        - df.repartition(n, col): shufffle, number partitions = min(n, unique record in col)
+    """
+
+    current_n_parts = df.rdd.getNumPartitions()
+
+    if n_files != None and repartition_by != None:
+        df_out = df.repartition(n_files, repartition_by)
+    elif n_files == None and repartition_by != None:
+        df_out = df.repartition(repartition_by)
+    elif n_files != None and repartition_by == None:
+        df_out = df.repartition(n_files)
+    else:
+        df_out = df
+
+    # for reporting out:
+    if repartition_by != None:
+        n_out_files = df.select(repartition_by).distinct().count() + 1
+        ext = f", repartitioned by {repartition_by}"
+    else:
+        n_out_files = current_n_parts
+        ext = ""
+    n_out_files = min(n_out_files, df_out.rdd.getNumPartitions())
+    logger.info(f"Saving {current_n_parts} partitions --> {n_out_files} file(s){ext}...")
+
+    df_out.write.mode("overwrite").option("path", filepath).saveAsTable(
+        "load_data_lookup", format="parquet"
+    )
+
+
+####################################################################
+# MAIN FUNCS
+@timed_info
+def run(relocated_file, lookup_file):
+    """ read from relocated_file, replace lookup_file with new output """
+
+    spark = init_spark_main("dsgrid-load")
+
+    # 1. load data
+    df_lookup = spark.read.parquet(relocated_file)
+
+    # 2. get keys to enumerte on
+    keys_to_exclude = ["scale_factor", "data_id", "id"]
+    keys = [x for x in df_lookup.columns if x not in keys_to_exclude]
+    logger.info(f"keys in load_data_lookup: {keys}")
+
+    # 3. enumerate keys
+    df_lookup_full = enumerate_lookup_by_keys(df_lookup, keys)
+
+    # 4. data quality check
+    enumeration_report(df_lookup_full, df_lookup)
+    assertion_checks(df_lookup_full, df_lookup, keys)
+
+    # 5. save
+    # 5.1. explore partitioning options
+    df_cols = df_lookup_full.columns
+
+    partition_stats = []
+    for key in df_cols:
+        report = file_size_if_partition_by(df_lookup_full, key)
+        partition_stats.append(report)
+
+    partition_stats = pd.concat(partition_stats, axis=1)
+
+    # 5.2. save to file by controling n_files
+    n_files = get_optimal_number_of_files(df_lookup_full)
+    save_file(df_lookup_full, lookup_file, n_files, repartition_by=None)
```

Comment:
How big is the load_data_lookup table after this process? Is there a log file you could share? (email or attach here)

Reply:
Still pretty small. 
[enumerate_load_table_output.log](https://github.com/dsgrid/dsgrid/files/7559086/enumerate_load_table_output.log)